File: app/src/main/java/rusha/x/recipe/list/RecipeListFragment.java
package rusha.x.recipe.list;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import javax.inject.Inject;

import rusha.x.MainActivity;
import rusha.x.databinding.RecipeListActivityBinding;

public class RecipeListFragment extends Fragment {
    @Inject
    public RecipeListViewModel viewModel;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MainActivity.di(this).inject(this);
    }

    private RecipeListActivityBinding binding;

    @Override
    public View onCreateView(
            LayoutInflater inflater,
            ViewGroup container,
            Bundle savedInstanceState
    ) {
        binding = RecipeListActivityBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        binding.addRecipeButton.setOnClickListener(v -> {
            viewModel.onAddRecipeClick();
        });

        binding.swipeRefreshLayout.setOnRefreshListener(() -> {
            viewModel.onRefresh();
        });

        viewModel.recipesLiveData.observe(getViewLifecycleOwner(), (recipes) -> {
            RecipesListAdapter recipesViewAdapter = new RecipesListAdapter(viewModel);
            recipesViewAdapter.setRecipesToAdopt(recipes);
            binding.recipesView.setAdapter(recipesViewAdapter);
        });

        viewModel.isRefreshingLiveData.observe(getViewLifecycleOwner(), (isRefreshing) -> {
            binding.swipeRefreshLayout.setRefreshing(isRefreshing);
        });
        viewModel.goToEditRecipe.observe(getViewLifecycleOwner(), (goToEditRecipe) -> {
            Navigation.findNavController(view).navigate(
                    RecipeListFragmentDirections.actionRecipeListFragmentToEditRecipeFragment()
            );
        });
        viewModel.goToRecipeDetails.observe(getViewLifecycleOwner(), (recipe) -> {
            Navigation.findNavController(view).navigate(
                    RecipeListFragmentDirections.actionRecipeListFragmentToRecipeDetailsFragment(recipe)
            );
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        viewModel.onResume();
    }
}

File: app/src/main/java/rusha/x/recipe/list/RecipeListFragment.kt
package rusha.x.recipe.list

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import rusha.x.R
import rusha.x.databinding.RecipeListActivityBinding
import rusha.x.di
import rusha.x.recipe.Recipe
import javax.inject.Inject

class RecipeListFragment : Fragment() {
    @Inject
    lateinit var viewModel: RecipeListViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        di().inject(this)
    }

    private lateinit var binding: RecipeListActivityBinding
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = RecipeListActivityBinding.inflate(inflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.addRecipeButton.setOnClickListener {
            viewModel.onAddRecipeClick()
        }

        binding.swipeRefreshLayout.setOnRefreshListener {
            viewModel.onRefresh()
        }

        viewModel.recipesLiveData.observe(viewLifecycleOwner, Observer { recipes ->
            val recipesViewAdapter = RecipesListAdapter(viewModel)
            recipesViewAdapter.recipesToAdopt = recipes
            binding.recipesView.adapter = recipesViewAdapter
        })

        viewModel.isRefreshingLiveData.observe(viewLifecycleOwner, Observer { isRefreshing ->
            binding.swipeRefreshLayout.isRefreshing = isRefreshing
        })
        viewModel.goToEditRecipe.observe(viewLifecycleOwner, Observer { goToEditRecipe ->
            findNavController().navigate(
                RecipeListFragmentDirections.actionRecipeListFragmentToEditRecipeFragment()
            )
        })
        viewModel.goToRecipeDetails.observe(viewLifecycleOwner, Observer { recipe: Recipe ->
            findNavController().navigate(
                RecipeListFragmentDirections.actionRecipeListFragmentToRecipeDetailsFragment(recipe)
            )
        })
    }

    override fun onResume() {
        super.onResume()
        viewModel.onResume()
    }
}
